codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +403,166 @@ public Http2Settings pollSentSettings() {
     public Configuration configuration() {
         return frameWriter.configuration();
     }
+
+    /**
+     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledBase {
+        private ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, padding, endOfStream, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            ReferenceCountUtil.safeRelease(data);
```

Comment:
I agree we don't want to have this method throw, or risk a recursive exception trap.  Instead propagate the exception up if we expect one might occur. (i.e. `cause = new RuntimeException("bad stuff", e);`.

The situation I think we are accounting for is for the "last" write if an exception occurs then we may be double releasing (because `writeData` always releases `data`)?  In this situation with the current code we would be double releasing and I see how we may not want to propagate the error.  However couldn't we have `if (data != EMPTY_BUFFER && data != null)` to protect against the double release?  My concern is if an exception is thrown during some other code path we should not discard the exception.